NSURLProtocol and NSURLSession do not guarantee delegate dispatch order and thread

My last CI build failed in the test execution phase; I've seen this several times over the past few months.

2017-04-14T17:18:29.7225794Z Non-passing Tests:
2017-04-14T17:18:29.7225794Z 
2017-04-14T17:18:29.7225794Z ##[error]:     NSURLSessionTests::DataTaskWithURL_WithCompletionHandler [Failed]
2017-04-14T17:18:29.7225794Z 
2017-04-14T17:18:29.7225794Z Summary: Total=112, Passed=111, Failed=1, Blocked=0, Not Run=0, Skipped=0
2017-04-14T17:18:29.8913424Z ##[error]Process completed with exit code 1.

--- a/Frameworks/Foundation/NSBlocks.mm
+++ b/Frameworks/Foundation/NSBlocks.mm
@@ -32,7 +32,6 @@
     Block_release(self);
 }
 
-// VSO 6196093: We are using this internally, but we should move away from it.
 - (void)invoke {
     void (^selfAsBlock)(void) = reinterpret_cast<void (^)(void)>(static_cast<void*>(self));
     selfAsBlock();

--- a/Frameworks/Foundation/NSURLConnection.mm
+++ b/Frameworks/Foundation/NSURLConnection.mm
@@ -126,18 +126,9 @@ static void __dispatchDelegateCallback(NSURLConnection* connection, void (^callb
     } else if (connection->_scheduledRunLoops.size() == 1) {
         // TODO #2469: Cannot currently schedule on more than one runloop
         for (auto pair : connection->_scheduledRunLoops) {
-            [pair.first performSelector:@selector(_invokeBlock:)
-                                 target:connection
-                               argument:callbackBlock
-                                  order:0
-                                  modes:@[ pair.second.get() ]];
+            [pair.first performSelector:@selector(invoke) target:callbackBlock argument:nil order:0 modes:@[ pair.second.get() ]];
         }
     }
-}
-
-// Helper function that just invokes a block
-- (void)_invokeBlock:(void (^)())block {
-    block();
 }
 
 // Body of the thread on which the NSURLProtocol loads

--- a/Frameworks/Foundation/NSURLConnection.mm
+++ b/Frameworks/Foundation/NSURLConnection.mm
@@ -382,19 +382,17 @@ static void __dispatchDelegateCallback(NSURLConnection* connection, void (^callb
  @Status Interoperable
 */
 - (void)URLProtocol:(NSURLProtocol*)urlProtocol didFailWithError:(NSError*)error {
-    __block StrongId<NSURLConnection> strongSelf = self;
-    __block StrongId<NSError> strongError = error;
     __dispatchDelegateCallback(self, ^void() {
         TraceError(TAG, L"URL protocol did fail");
         // if ( [_delegate respondsToSelector:@selector(connection:willSendRequest:redirectResponse:)] ) [_delegate
         // connection:self willSendRequest:_request redirectResponse:nil];
-        if ([strongSelf->_delegate respondsToSelector:@selector(connection:didFailWithError:)]) {
-            [strongSelf->_delegate connection:strongSelf didFailWithError:strongError];
+        if ([_delegate respondsToSelector:@selector(connection:didFailWithError:)]) {
+            [_delegate connection:self didFailWithError:error];
         }
 
-        strongSelf->_protocol = nil;
-        strongSelf->_delegate = nil;
-        strongSelf->_done = true;
+        _protocol = nil;
+        _delegate = nil;
+        _done = true;
     });
 }
 
@@ -404,8 +402,6 @@ static void __dispatchDelegateCallback(NSURLConnection* connection, void (^callb
 - (void)URLProtocol:(NSURLProtocol*)urlProtocol
     didReceiveResponse:(NSURLResponse*)response
     cacheStoragePolicy:(NSURLCacheStoragePolicy)policy {
-    __block StrongId<NSURLConnection> strongSelf = self;
-    __block StrongId<NSURLResponse> strongResponse = response;
     __dispatchDelegateCallback(self, ^void() {
         /*
         if ( [response respondsToSelector:@selector(statusCode)] && [response statusCode] != 200 ) {
@@ -413,15 +409,15 @@ static void __dispatchDelegateCallback(NSURLConnection* connection, void (^callb
         }
         */
 
-        strongSelf->_response = response;
-        strongSelf->_storagePolicy = policy;
+        _response = response;
+        _storagePolicy = policy;
 
-        if ([strongSelf->_delegate respondsToSelector:@selector(connection:willCacheResponse:)]) {
+        if ([_delegate respondsToSelector:@selector(connection:willCacheResponse:)]) {
             // TODO #2469: This is currently incorrect, these are not the same thing
-            [strongSelf->_delegate connection:strongSelf willCacheResponse:static_cast<NSCachedURLResponse*>(response)];
+            [_delegate connection:self willCacheResponse:static_cast<NSCachedURLResponse*>(response)];
         }
-        if ([strongSelf->_delegate respondsToSelector:@selector(connection:didReceiveResponse:)]) {
-            [strongSelf->_delegate connection:strongSelf didReceiveResponse:strongResponse];
+        if ([_delegate respondsToSelector:@selector(connection:didReceiveResponse:)]) {
+            [_delegate connection:self didReceiveResponse:response];
         }
     });
 }
@@ -430,11 +426,9 @@ static void __dispatchDelegateCallback(NSURLConnection* connection, void (^callb
  @Status Interoperable
 */
 - (void)URLProtocol:(NSURLProtocol*)urlProtocol didLoadData:(NSData*)data {
-    __block StrongId<NSURLConnection> strongSelf = self;
-    __block StrongId<NSData*> strongData = data;
     __dispatchDelegateCallback(self, ^void() {
-        if ([strongSelf->_delegate respondsToSelector:@selector(connection:didReceiveData:)]) {
-            [strongSelf->_delegate connection:strongSelf didReceiveData:strongData];
+        if ([_delegate respondsToSelector:@selector(connection:didReceiveData:)]) {
+            [_delegate connection:self didReceiveData:data];
         }
     });
 }
@@ -443,7 +437,6 @@ static void __dispatchDelegateCallback(NSURLConnection* connection, void (^callb
  @Status Interoperable
 */
 - (void)URLProtocolDidFinishLoading:(NSURLProtocol*)urlProtocol {
-    __block StrongId<NSURLConnection> strongSelf = self;
     __dispatchDelegateCallback(self, ^void() {
         /*
         if(_storagePolicy==NSURLCacheStorageNotAllowed) {
@@ -461,13 +454,13 @@ static void __dispatchDelegateCallback(NSURLConnection* connection, void (^callb
         }
         */
 
-        if ([strongSelf->_delegate respondsToSelector:@selector(connectionDidFinishLoading:)]) {
-            [strongSelf->_delegate connectionDidFinishLoading:strongSelf];
+        if ([_delegate respondsToSelector:@selector(connectionDidFinishLoading:)]) {
+            [_delegate connectionDidFinishLoading:self];
         }
 
-        strongSelf->_protocol = nil;
-        strongSelf->_delegate = nil;
-        strongSelf->_done = true;
+        _protocol = nil;
+        _delegate = nil;
+        _done = true;
     });
 }
 
@@ -475,13 +468,11 @@ static void __dispatchDelegateCallback(NSURLConnection* connection, void (^callb
  @Status Interoperable
 */
 - (void)URLProtocol:(NSURLProtocol*)urlProtocol didReceiveAuthenticationChallenge:(NSURLAuthenticationChallenge*)challenge {
-    __block StrongId<NSURLConnection> strongSelf = self;
-    __block StrongId<NSURLAuthenticationChallenge> strongChallenge = challenge;
     __dispatchDelegateCallback(self, ^void() {
-        if ([strongSelf->_delegate respondsToSelector:@selector(connection:willSendRequestForAuthenticationChallenge:)]) {
-            [strongSelf->_delegate connection:strongSelf willSendRequestForAuthenticationChallenge:strongChallenge];
+        if ([_delegate respondsToSelector:@selector(connection:willSendRequestForAuthenticationChallenge:)]) {
+            [_delegate connection:self willSendRequestForAuthenticationChallenge:challenge];
         } else {
-            [strongSelf->_delegate connection:strongSelf didReceiveAuthenticationChallenge:strongChallenge];
+            [_delegate connection:self didReceiveAuthenticationChallenge:challenge];
         }
     });
 }
@@ -490,34 +481,28 @@ static void __dispatchDelegateCallback(NSURLConnection* connection, void (^callb
  @Status Interoperable
 */
 - (void)URLProtocol:(NSURLProtocol*)urlProtocol wasRedirectedToRequest:(NSURLRequest*)request redirectResponse:(NSURLResponse*)response {
-    __block StrongId<NSURLConnection> strongSelf = self;
-    __block StrongId<NSURLRequest> strongRequest = request;
-    __block StrongId<NSURLResponse> strongResponse = response;
     __dispatchDelegateCallback(self, ^void() {
-        [strongSelf->_protocol stopLoading];
+        [_protocol stopLoading];
         NSURLRequest* newRequest = request;
-        if ([strongSelf->_delegate respondsToSelector:@selector(connection:willSendRequest:redirectResponse:)]) {
-            newRequest = [strongSelf->_delegate connection:strongSelf willSendRequest:strongRequest redirectResponse:strongResponse];
+        if ([_delegate respondsToSelector:@selector(connection:willSendRequest:redirectResponse:)]) {
+            newRequest = [_delegate connection:self willSendRequest:request redirectResponse:response];
         }
-        strongSelf->_protocol = nil;
+        _protocol = nil;
 
         if (!newRequest) {
-            [strongSelf cancel];
+            [self cancel];
             return;
         }
 
-        NSURLProtocol* protocol = __protocolForRequest(request, strongSelf.get());
+        NSURLProtocol* protocol = __protocolForRequest(request, self);
         if (!protocol) {
             TraceError(TAG, L"NSURLConnection was redirected to a new request, but could not create a NSURLProtocol for the new request.");
             return;
         }
 
-        strongSelf->_protocol = protocol;
-        strongSelf->_currentRequest = newRequest;
-        [strongSelf->_protocol performSelector:@selector(startLoading)
-                                      onThread:strongSelf->_protocolThread
-                                    withObject:nil
-                                 waitUntilDone:NO];
+        _protocol = protocol;
+        _currentRequest = newRequest;
+        [_protocol performSelector:@selector(startLoading) onThread:_protocolThread withObject:nil waitUntilDone:NO];
     });
 }
 

--- a/Frameworks/Foundation/NSURLConnection.mm
+++ b/Frameworks/Foundation/NSURLConnection.mm
@@ -24,6 +24,8 @@ OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 #import "LoggingNative.h"
 #import "NSURLProtocolInternal.h"
 
+#import <NSRunLoopSource.h>
+
 #import <algorithm>
 #import <vector>
 
@@ -110,7 +112,41 @@ static NSURLProtocol* __protocolForRequest(NSURLRequest* request, id<NSURLProtoc
     NSURLCacheStoragePolicy _storagePolicy;
     StrongId<NSURLProtocol> _protocol;
 
+    StrongId<NSRunLoopSource> _runLoopCancelSource;
+    StrongId<NSThread> _protocolThread;
     bool _started;
+    bool _done;
+}
+
+// Helper function that executes a block on the delegate queue or scheduled run loops - used for delegate callbacks
+static void __dispatchDelegateCallback(NSURLConnection* connection, void (^callbackBlock)()) {
+    if (connection->_delegateQueue) {
+        [connection->_delegateQueue addOperationWithBlock:callbackBlock];
+
+    } else if (connection->_scheduledRunLoops.size() == 1) {
+        // TODO #2469: Cannot currently schedule on more than one runloop
+        for (auto pair : connection->_scheduledRunLoops) {
+            [pair.first performSelector:@selector(_invokeBlock:)
+                                 target:connection
+                               argument:callbackBlock
+                                  order:0
+                                  modes:@[ pair.second.get() ]];
+        }
+    }
+}
+
+// Helper function that just invokes a block
+- (void)_invokeBlock:(void (^)())block {
+    block();
+}
+
+// Body of the thread on which the NSURLProtocol loads
+- (void)_protocolThreadBody {
+    NSRunLoop* currentRunLoop = [NSRunLoop currentRunLoop];
+    [currentRunLoop _addInputSource:_runLoopCancelSource forMode:NSDefaultRunLoopMode];
+    while (!_done) {
+        [currentRunLoop runUntilDate:[NSDate distantFuture]];
+    }
 }
 
 /**
@@ -195,6 +231,7 @@ static NSURLProtocol* __protocolForRequest(NSURLRequest* request, id<NSURLProtoc
         _originalRequest = [request copy];
         _currentRequest = [_originalRequest copy];
         _delegate = delegate;
+        _runLoopCancelSource.attach([NSRunLoopSource new]);
 
         if (startImmediately) {
             [self start];
@@ -237,6 +274,11 @@ static NSURLProtocol* __protocolForRequest(NSURLRequest* request, id<NSURLProtoc
             return;
         }
 
+        if ((!_delegateQueue) && (_scheduledRunLoops.empty())) {
+            // Run on the current run loop in the default mode
+            [self scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];
+        }
+
         _started = true;
 
         NSURLProtocol* protocol = __protocolForRequest(_currentRequest, self);
@@ -246,25 +288,10 @@ static NSURLProtocol* __protocolForRequest(NSURLRequest* request, id<NSURLProtoc
         }
         _protocol = protocol;
 
-        if (_delegateQueue) {
-            [_delegateQueue addOperationWithBlock:^void() {
-                [_protocol startLoading];
-            }];
+        _protocolThread.attach([[NSThread alloc] initWithTarget:self selector:@selector(_protocolThreadBody) object:nil]);
+        [_protocolThread start];
 
-        } else if (_scheduledRunLoops.size() == 1) {
-            // TODO #2469: Cannot currently schedule on more than one runloop
-            for (auto pair : _scheduledRunLoops) {
-                [pair.first performSelector:@selector(startLoading) target:_protocol argument:nil order:0 modes:@[ pair.second.get() ]];
-            }
-
-        } else {
-            // Run on the current run loop in the default mode
-            [[NSRunLoop currentRunLoop] performSelector:@selector(startLoading)
-                                                 target:_protocol
-                                               argument:nil
-                                                  order:0
-                                                  modes:@[ NSDefaultRunLoopMode ]];
-        }
+        [_protocol performSelector:@selector(startLoading) onThread:_protocolThread withObject:nil waitUntilDone:NO];
     }
 }
 
@@ -346,118 +373,152 @@ static NSURLProtocol* __protocolForRequest(NSURLRequest* request, id<NSURLProtoc
     @synchronized(self) {
         [_protocol stopLoading];
         _protocol = nil;
+
+        [_runLoopCancelSource _trigger];
     }
 }
 
 /**
  @Status Interoperable
 */
-- (void)URLProtocol:(id)urlProtocol didFailWithError:(id)error {
-    TraceError(TAG, L"URL protocol did fail");
-    // if ( [_delegate respondsToSelector:@selector(connection:willSendRequest:redirectResponse:)] ) [_delegate
-    // connection:self willSendRequest:_request redirectResponse:nil];
-    if ([_delegate respondsToSelector:@selector(connection:didFailWithError:)]) {
-        [_delegate connection:self didFailWithError:error];
-    }
+- (void)URLProtocol:(NSURLProtocol*)urlProtocol didFailWithError:(NSError*)error {
+    __block StrongId<NSURLConnection> strongSelf = self;
+    __block StrongId<NSError> strongError = error;
+    __dispatchDelegateCallback(self, ^void() {
+        TraceError(TAG, L"URL protocol did fail");
+        // if ( [_delegate respondsToSelector:@selector(connection:willSendRequest:redirectResponse:)] ) [_delegate
+        // connection:self willSendRequest:_request redirectResponse:nil];
+        if ([strongSelf->_delegate respondsToSelector:@selector(connection:didFailWithError:)]) {
+            [strongSelf->_delegate connection:strongSelf didFailWithError:strongError];
+        }
 
-    _protocol = nil;
-    _delegate = nil;
+        strongSelf->_protocol = nil;
+        strongSelf->_delegate = nil;
+        strongSelf->_done = true;
+    });
 }
 
 /**
  @Status Interoperable
 */
-- (void)URLProtocol:(id)urlProtocol didReceiveResponse:(NSURLResponse*)response cacheStoragePolicy:(NSURLCacheStoragePolicy)policy {
-    /*
-    if ( [response respondsToSelector:@selector(statusCode)] && [response statusCode] != 200 ) {
-    [_delegate setError:[NSError errorWithDomain:@"Bad response code" code:[response statusCode] userInfo:nil]];
-    }
-    */
+- (void)URLProtocol:(NSURLProtocol*)urlProtocol
+    didReceiveResponse:(NSURLResponse*)response
+    cacheStoragePolicy:(NSURLCacheStoragePolicy)policy {
+    __block StrongId<NSURLConnection> strongSelf = self;
+    __block StrongId<NSURLResponse> strongResponse = response;
+    __dispatchDelegateCallback(self, ^void() {
+        /*
+        if ( [response respondsToSelector:@selector(statusCode)] && [response statusCode] != 200 ) {
+        [_delegate setError:[NSError errorWithDomain:@"Bad response code" code:[response statusCode] userInfo:nil]];
+        }
+        */
 
-    _response = response;
-    _storagePolicy = policy;
+        strongSelf->_response = response;
+        strongSelf->_storagePolicy = policy;
 
-    if ([_delegate respondsToSelector:@selector(connection:willCacheResponse:)]) {
-        // TODO #2469: This is currently incorrect, these are not the same thing
-        [_delegate connection:self willCacheResponse:static_cast<NSCachedURLResponse*>(response)];
-    }
-    if ([_delegate respondsToSelector:@selector(connection:didReceiveResponse:)]) {
-        [_delegate connection:self didReceiveResponse:response];
-    }
+        if ([strongSelf->_delegate respondsToSelector:@selector(connection:willCacheResponse:)]) {
+            // TODO #2469: This is currently incorrect, these are not the same thing
+            [strongSelf->_delegate connection:strongSelf willCacheResponse:static_cast<NSCachedURLResponse*>(response)];
+        }
+        if ([strongSelf->_delegate respondsToSelector:@selector(connection:didReceiveResponse:)]) {
+            [strongSelf->_delegate connection:strongSelf didReceiveResponse:strongResponse];
+        }
+    });
 }
 
 /**
  @Status Interoperable
 */
-- (void)URLProtocol:(id)urlProtocol didLoadData:(id)data {
-    if ([_delegate respondsToSelector:@selector(connection:didReceiveData:)]) {
-        [_delegate connection:self didReceiveData:data];
-    }
+- (void)URLProtocol:(NSURLProtocol*)urlProtocol didLoadData:(NSData*)data {
+    __block StrongId<NSURLConnection> strongSelf = self;
+    __block StrongId<NSData*> strongData = data;
+    __dispatchDelegateCallback(self, ^void() {
+        if ([strongSelf->_delegate respondsToSelector:@selector(connection:didReceiveData:)]) {
+            [strongSelf->_delegate connection:strongSelf didReceiveData:strongData];
+        }
+    });
 }
 
 /**
  @Status Interoperable
 */
-- (void)URLProtocolDidFinishLoading:(id)urlProtocol {
-    /*
-    if(_storagePolicy==NSURLCacheStorageNotAllowed) {
-    //[[NSURLCache sharedURLCache] removeCachedResponseForRequest:_request];
-    } else {
-    //NSCachedURLResponse *cachedResponse=[[NSCachedURLResponse alloc] initWithResponse:_response data:_mutableData
-    userInfo:nil storagePolicy:_storagePolicy];
+- (void)URLProtocolDidFinishLoading:(NSURLProtocol*)urlProtocol {
+    __block StrongId<NSURLConnection> strongSelf = self;
+    __dispatchDelegateCallback(self, ^void() {
+        /*
+        if(_storagePolicy==NSURLCacheStorageNotAllowed) {
+        //[[NSURLCache sharedURLCache] removeCachedResponseForRequest:_request];
+        } else {
+        //NSCachedURLResponse *cachedResponse=[[NSCachedURLResponse alloc] initWithResponse:_response data:_mutableData
+        userInfo:nil storagePolicy:_storagePolicy];
 
-    //if([_delegate respondsToSelector:@selector(connection:willCacheResponse:)])
-    //cachedResponse=[_delegate connection:self willCacheResponse:cachedResponse];
+        //if([_delegate respondsToSelector:@selector(connection:willCacheResponse:)])
+        //cachedResponse=[_delegate connection:self willCacheResponse:cachedResponse];
 
-    //if(cachedResponse!=nil){
-    //[[NSURLCache sharedURLCache] storeCachedResponse:cachedResponse forRequest:_request];
-    //}
-    }
-    */
+        //if(cachedResponse!=nil){
+        //[[NSURLCache sharedURLCache] storeCachedResponse:cachedResponse forRequest:_request];
+        //}
+        }
+        */
 
-    if ([_delegate respondsToSelector:@selector(connectionDidFinishLoading:)]) {
-        [_delegate performSelector:@selector(connectionDidFinishLoading:) withObject:self];
-    }
+        if ([strongSelf->_delegate respondsToSelector:@selector(connectionDidFinishLoading:)]) {
+            [strongSelf->_delegate connectionDidFinishLoading:strongSelf];
+        }
 
-    _protocol = nil;
-    _delegate = nil;
+        strongSelf->_protocol = nil;
+        strongSelf->_delegate = nil;
+        strongSelf->_done = true;
+    });
 }
 
 /**
  @Status Interoperable
 */
-- (void)URLProtocol:(id)urlProtocol didReceiveAuthenticationChallenge:(id)challenge {
-    if ([_delegate respondsToSelector:@selector(connection:willSendRequestForAuthenticationChallenge:)]) {
-        [_delegate connection:self willSendRequestForAuthenticationChallenge:challenge];
-    } else {
-        [_delegate connection:self didReceiveAuthenticationChallenge:challenge];
-    }
+- (void)URLProtocol:(NSURLProtocol*)urlProtocol didReceiveAuthenticationChallenge:(NSURLAuthenticationChallenge*)challenge {
+    __block StrongId<NSURLConnection> strongSelf = self;
+    __block StrongId<NSURLAuthenticationChallenge> strongChallenge = challenge;
+    __dispatchDelegateCallback(self, ^void() {
+        if ([strongSelf->_delegate respondsToSelector:@selector(connection:willSendRequestForAuthenticationChallenge:)]) {
+            [strongSelf->_delegate connection:strongSelf willSendRequestForAuthenticationChallenge:strongChallenge];
+        } else {
+            [strongSelf->_delegate connection:strongSelf didReceiveAuthenticationChallenge:strongChallenge];
+        }
+    });
 }
 
 /**
  @Status Interoperable
 */
 - (void)URLProtocol:(NSURLProtocol*)urlProtocol wasRedirectedToRequest:(NSURLRequest*)request redirectResponse:(NSURLResponse*)response {
-    [_protocol stopLoading];
-    NSURLRequest* newRequest = request;
-    if ([_delegate respondsToSelector:@selector(connection:willSendRequest:redirectResponse:)]) {
-        newRequest = [_delegate connection:self willSendRequest:request redirectResponse:response];
-    }
-    _protocol = nil;
+    __block StrongId<NSURLConnection> strongSelf = self;
+    __block StrongId<NSURLRequest> strongRequest = request;
+    __block StrongId<NSURLResponse> strongResponse = response;
+    __dispatchDelegateCallback(self, ^void() {
+        [strongSelf->_protocol stopLoading];
+        NSURLRequest* newRequest = request;
+        if ([strongSelf->_delegate respondsToSelector:@selector(connection:willSendRequest:redirectResponse:)]) {
+            newRequest = [strongSelf->_delegate connection:strongSelf willSendRequest:strongRequest redirectResponse:strongResponse];
+        }
+        strongSelf->_protocol = nil;
 
-    if (!newRequest) {
-        [self cancel];
-        return;
-    }
+        if (!newRequest) {
+            [strongSelf cancel];
+            return;
+        }
 
-    NSURLProtocol* protocol = __protocolForRequest(request, self);
-    if (!protocol) {
-        TraceError(TAG, L"NSURLConnection was redirected to a new request, but could not create a NSURLProtocol for the new request.");
-        return;
-    }
-    _protocol = protocol;
-    _currentRequest = newRequest;
-    [_protocol startLoading];
+        NSURLProtocol* protocol = __protocolForRequest(request, strongSelf.get());
+        if (!protocol) {
+            TraceError(TAG, L"NSURLConnection was redirected to a new request, but could not create a NSURLProtocol for the new request.");
+            return;
+        }
+
+        strongSelf->_protocol = protocol;
+        strongSelf->_currentRequest = newRequest;
+        [strongSelf->_protocol performSelector:@selector(startLoading)
+                                      onThread:strongSelf->_protocolThread
+                                    withObject:nil
+                                 waitUntilDone:NO];
+    });
 }
 
 /**

--- a/Frameworks/Foundation/NSURLProtocol.mm
+++ b/Frameworks/Foundation/NSURLProtocol.mm
@@ -81,14 +81,6 @@ static const wchar_t* TAG = L"NSURLProtocol";
     return self;
 }
 
-- (id)scheduleInRunLoop:(id)runLoop forMode:(id)mode {
-    return self;
-}
-
-- (id)unscheduleFromRunLoop:(id)runLoop forMode:(id)mode {
-    return self;
-}
-
 /**
  @Status Interoperable
 */

--- a/Frameworks/Foundation/NSURLProtocol_WinHTTP.mm
+++ b/Frameworks/Foundation/NSURLProtocol_WinHTTP.mm
@@ -247,15 +247,10 @@ static std::function<R(Args...)> bindObjC(id instance, SEL _cmd) {
 // Helper function that executes a block on the client thread - used for client callbacks
 static void __dispatchClientCallback(NSURLProtocol_WinHTTP* protocol, void (^callbackBlock)()) {
     if (protocol->_clientThread) {
-        [protocol performSelector:@selector(_invokeBlock:) onThread:protocol->_clientThread withObject:callbackBlock waitUntilDone:NO];
+        [callbackBlock performSelector:@selector(invoke) onThread:protocol->_clientThread withObject:nil waitUntilDone:NO];
     } else {
         callbackBlock();
     }
-}
-
-// Helper function that just invokes a block
-- (void)_invokeBlock:(void (^)())block {
-    block();
 }
 
 - (void)startLoading {
@@ -411,10 +406,10 @@ static void __dispatchClientCallback(NSURLProtocol_WinHTTP* protocol, void (^cal
                 default:
                     httpVersionStr = nil;
             }
-            __block StrongId<NSHTTPURLResponse> nsResponse = [[[NSHTTPURLResponse alloc] initWithURL:[request URL]
-                                                                                          statusCode:statusCode
-                                                                                         HTTPVersion:httpVersionStr
-                                                                                        headerFields:nsHeaders] autorelease];
+            NSHTTPURLResponse* nsResponse = [[[NSHTTPURLResponse alloc] initWithURL:[request URL]
+                                                                         statusCode:statusCode
+                                                                        HTTPVersion:httpVersionStr
+                                                                       headerFields:nsHeaders] autorelease];
 
             if (statusCode >= 300 && statusCode <= 399) { // redirect code
                 ComPtr<IUriRuntimeClass> uri;
@@ -428,7 +423,7 @@ static void __dispatchClientCallback(NSURLProtocol_WinHTTP* protocol, void (^cal
 
                 // The location header can specify a relative or absolute URL: attempt to join it with the request URL.
                 NSURL* targetUrl = [NSURL URLWithString:Strings::WideToNSString(uriString.Get()) relativeToURL:[request URL]];
-                __block StrongId<NSURLRequest> req = [NSURLRequest requestWithURL:targetUrl];
+                NSURLRequest* req = [NSURLRequest requestWithURL:targetUrl];
 
                 __dispatchClientCallback(self, ^void() {
                     [self.client URLProtocol:self wasRedirectedToRequest:req redirectResponse:nsResponse];
@@ -462,15 +457,13 @@ static void __dispatchClientCallback(NSURLProtocol_WinHTTP* protocol, void (^cal
 }
 
 - (void)_propagateErrorToClient:(NSError*)error {
-    __block StrongId<NSError> strongError = error;
     __dispatchClientCallback(self, ^void() {
-        [self.client URLProtocol:self didFailWithError:strongError];
+        [self.client URLProtocol:self didFailWithError:error];
     });
 }
 
 - (void)_consumeDataStreamForIHttpContent:(IHttpContent*)pHttpContent {
     ComPtr<IHttpContent> httpContent(pHttpContent);
-    __block StrongId<NSURLProtocol_WinHTTP> strongSelf = self;
     NSError* error = nil;
     try {
         ComPtr<IInputStream> stream;
@@ -486,7 +479,7 @@ static void __dispatchClientCallback(NSURLProtocol_WinHTTP* protocol, void (^cal
         ComPtr<IBuffer> outputBuffer;
 
         while (true) {
-            if ([strongSelf cancelled]) {
+            if ([self cancelled]) {
                 return;
             }
 
@@ -497,7 +490,7 @@ static void __dispatchClientCallback(NSURLProtocol_WinHTTP* protocol, void (^cal
                 outputBuffer.ReleaseAndGetAddressOf());
             THROW_IF_FAILED(hr);
 
-            if ([strongSelf cancelled]) { // double check on cancel after a long-running async operation.
+            if ([self cancelled]) { // double check on cancel after a long-running async operation.
                 return;
             }
 
@@ -515,20 +508,20 @@ static void __dispatchClientCallback(NSURLProtocol_WinHTTP* protocol, void (^cal
             // We're opting to take a copy of the buffer here, as NSURLProtocol's consumer can legally expect the data
             // to remain valid long after the connection is gone. Eventually, it would be very nice to have an IBuffer-backed
             // NSData.
-            __block StrongId<NSData> data = [NSData dataWithBytes:pOutputBuffer length:length];
+            NSData* data = [NSData dataWithBytes:pOutputBuffer length:length];
             __dispatchClientCallback(self, ^void() {
-                [self.client URLProtocol:strongSelf didLoadData:data];
+                [self.client URLProtocol:self didLoadData:data];
             });
         }
 
         __dispatchClientCallback(self, ^void() {
-            [self.client URLProtocolDidFinishLoading:strongSelf];
+            [self.client URLProtocolDidFinishLoading:self];
         });
     }
     CATCH_POPULATE_NSERROR(&error);
 
     if (error) {
-        [strongSelf _propagateErrorToClient:error];
+        [self _propagateErrorToClient:error];
     }
 }
 @end

--- a/Frameworks/Foundation/NSURLProtocol_WinHTTP.mm
+++ b/Frameworks/Foundation/NSURLProtocol_WinHTTP.mm
@@ -443,7 +443,7 @@ static void __dispatchClientCallback(NSURLProtocol_WinHTTP* protocol, void (^cal
 
             if (content) {
                 // we could be cancelled mid-operation, but we need to retain the ability to clean up.
-                __block StrongId<NSURLProtocol_WinHTTP> strongSelf = self;
+                StrongId<NSURLProtocol_WinHTTP> strongSelf = self;
                 dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
                     [strongSelf _consumeDataStreamForIHttpContent:content.Get()];
                 });

--- a/Frameworks/Foundation/NSURLProtocol_WinHTTP.mm
+++ b/Frameworks/Foundation/NSURLProtocol_WinHTTP.mm
@@ -211,6 +211,7 @@ static ComPtr<IHttpRequestMessage> _requestMessageForNS(NSURLRequest* nsRequest)
 @interface NSURLProtocol_WinHTTP : NSURLProtocol {
     ComPtr<AsyncHttpOperation> _httpRequestOperation;
     ComPtr<IHttpClient> _httpClient;
+    StrongId<NSThread> _clientThread;
 }
 @property (assign, atomic) bool cancelled;
 @end
@@ -243,20 +244,35 @@ static std::function<R(Args...)> bindObjC(id instance, SEL _cmd) {
     };
 }
 
+// Helper function that executes a block on the client thread - used for client callbacks
+static void __dispatchClientCallback(NSURLProtocol_WinHTTP* protocol, void (^callbackBlock)()) {
+    if (protocol->_clientThread) {
+        [protocol performSelector:@selector(_invokeBlock:) onThread:protocol->_clientThread withObject:callbackBlock waitUntilDone:NO];
+    } else {
+        callbackBlock();
+    }
+}
+
+// Helper function that just invokes a block
+- (void)_invokeBlock:(void (^)())block {
+    block();
+}
+
 - (void)startLoading {
     @synchronized(self) {
         if (_httpRequestOperation) {
             return;
         }
 
+        _clientThread = [NSThread currentThread];
+
         NSError* error = nil;
+
         try {
             auto httpRequestMessage = _requestMessageForNS(self.request);
             THROW_IF_FAILED(_httpClient->SendRequestWithOptionAsync(httpRequestMessage.Get(),
                                                                     HttpCompletionOption::HttpCompletionOption_ResponseHeadersRead,
                                                                     &_httpRequestOperation));
-
-            StrongId<NSURLProtocol_WinHTTP> strongSelf(self);
             auto handler = MakeHttpCompletedCallback(
                 bindObjC<HRESULT, AsyncHttpOperation*, AsyncStatus>(self, @selector(_asyncHttpOperation:completedWithStatus:)));
             THROW_IF_FAILED(_httpRequestOperation->put_Completed(handler.Get()));
@@ -395,10 +411,10 @@ static std::function<R(Args...)> bindObjC(id instance, SEL _cmd) {
                 default:
                     httpVersionStr = nil;
             }
-            NSHTTPURLResponse* nsResponse = [[[NSHTTPURLResponse alloc] initWithURL:[request URL]
-                                                                         statusCode:statusCode
-                                                                        HTTPVersion:httpVersionStr
-                                                                       headerFields:nsHeaders] autorelease];
+            __block StrongId<NSHTTPURLResponse> nsResponse = [[[NSHTTPURLResponse alloc] initWithURL:[request URL]
+                                                                                          statusCode:statusCode
+                                                                                         HTTPVersion:httpVersionStr
+                                                                                        headerFields:nsHeaders] autorelease];
 
             if (statusCode >= 300 && statusCode <= 399) { // redirect code
                 ComPtr<IUriRuntimeClass> uri;
@@ -412,19 +428,24 @@ static std::function<R(Args...)> bindObjC(id instance, SEL _cmd) {
 
                 // The location header can specify a relative or absolute URL: attempt to join it with the request URL.
                 NSURL* targetUrl = [NSURL URLWithString:Strings::WideToNSString(uriString.Get()) relativeToURL:[request URL]];
-                NSURLRequest* req = [NSURLRequest requestWithURL:targetUrl];
-                [self.client URLProtocol:self wasRedirectedToRequest:req redirectResponse:nsResponse];
+                __block StrongId<NSURLRequest> req = [NSURLRequest requestWithURL:targetUrl];
+
+                __dispatchClientCallback(self, ^void() {
+                    [self.client URLProtocol:self wasRedirectedToRequest:req redirectResponse:nsResponse];
+                });
                 // After a redirect, NSURLProtocol ceases to exist: we could redirect to a completely different one.
                 return S_OK;
             } else {
-                [self.client URLProtocol:self didReceiveResponse:nsResponse cacheStoragePolicy:NSURLCacheStorageNotAllowed];
+                __dispatchClientCallback(self, ^void() {
+                    [self.client URLProtocol:self didReceiveResponse:nsResponse cacheStoragePolicy:NSURLCacheStorageNotAllowed];
+                });
             }
 
             if (content) {
-                [self retain]; // we could be cancelled mid-operation, but we need to retain the ability to clean up.
+                // we could be cancelled mid-operation, but we need to retain the ability to clean up.
+                __block StrongId<NSURLProtocol_WinHTTP> strongSelf = self;
                 dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
-                    [self _consumeDataStreamForIHttpContent:content.Get()];
-                    [self release];
+                    [strongSelf _consumeDataStreamForIHttpContent:content.Get()];
                 });
             }
 
@@ -441,16 +462,15 @@ static std::function<R(Args...)> bindObjC(id instance, SEL _cmd) {
 }
 
 - (void)_propagateErrorToClient:(NSError*)error {
-    [self.client URLProtocol:self didFailWithError:error];
-}
-
-- (void)_didFinishLoading {
-    [self.client URLProtocolDidFinishLoading:self];
+    __block StrongId<NSError> strongError = error;
+    __dispatchClientCallback(self, ^void() {
+        [self.client URLProtocol:self didFailWithError:strongError];
+    });
 }
 
 - (void)_consumeDataStreamForIHttpContent:(IHttpContent*)pHttpContent {
     ComPtr<IHttpContent> httpContent(pHttpContent);
-    id strongSelf = self;
+    __block StrongId<NSURLProtocol_WinHTTP> strongSelf = self;
     NSError* error = nil;
     try {
         ComPtr<IInputStream> stream;
@@ -495,10 +515,15 @@ static std::function<R(Args...)> bindObjC(id instance, SEL _cmd) {
             // We're opting to take a copy of the buffer here, as NSURLProtocol's consumer can legally expect the data
             // to remain valid long after the connection is gone. Eventually, it would be very nice to have an IBuffer-backed
             // NSData.
-            [self.client URLProtocol:strongSelf didLoadData:[NSData dataWithBytes:pOutputBuffer length:length]];
+            __block StrongId<NSData> data = [NSData dataWithBytes:pOutputBuffer length:length];
+            __dispatchClientCallback(self, ^void() {
+                [self.client URLProtocol:strongSelf didLoadData:data];
+            });
         }
 
-        [strongSelf _didFinishLoading];
+        __dispatchClientCallback(self, ^void() {
+            [self.client URLProtocolDidFinishLoading:strongSelf];
+        });
     }
     CATCH_POPULATE_NSERROR(&error);
 

--- a/Frameworks/Foundation/NSURLProtocol_file.h
+++ b/Frameworks/Foundation/NSURLProtocol_file.h
@@ -21,8 +21,6 @@
 - (id)startLoading;
 - (id)stopLoading;
 - (id)statusVersion:(id)versionStr;
-- (id)scheduleInRunLoop:(id)runLoop forMode:(id)mode;
 - (id)_doFileLoad;
-- (id)unscheduleFromRunLoop:(id)runLoop forMode:(id)mode;
 + (BOOL)canInitWithRequest:(id)request;
 @end

--- a/Frameworks/Foundation/NSURLProtocol_file.mm
+++ b/Frameworks/Foundation/NSURLProtocol_file.mm
@@ -76,11 +76,6 @@ static const wchar_t* TAG = L"NSURLProtocol_file";
     return self;
 }
 
-- (id)scheduleInRunLoop:(id)runLoop forMode:(id)mode {
-    [runLoop performSelector:@selector(_doFileLoad) target:self argument:nil order:0 modes:[NSArray arrayWithObject:mode]];
-    return self;
-}
-
 - (id)_doFileLoad {
     NSData* dataReceived = [NSData dataWithContentsOfFile:_path];
 
@@ -89,10 +84,6 @@ static const wchar_t* TAG = L"NSURLProtocol_file";
     [client URLProtocol:self didLoadData:dataReceived];
     [client URLProtocolDidFinishLoading:self];
 
-    return self;
-}
-
-- (id)unscheduleFromRunLoop:(id)runLoop forMode:(id)mode {
     return self;
 }
 

--- a/tests/functionaltests/Tests/NSURLConnection.mm
+++ b/tests/functionaltests/Tests/NSURLConnection.mm
@@ -114,6 +114,21 @@ typedef NS_ENUM(NSInteger, NSURLConnectionDelegateType) {
     return connection;
 }
 
+- (void)connection:(NSURLConnection*)connection didReceiveResponse:(NSURLResponse*)response {
+    EXPECT_TRUE([_operationQueue isEqual:[NSOperationQueue currentQueue]]);
+    [super connection:connection didReceiveResponse:response];
+}
+
+- (void)connection:(NSURLConnection*)connection didReceiveData:(nonnull NSData*)data {
+    EXPECT_TRUE([_operationQueue isEqual:[NSOperationQueue currentQueue]]);
+    [super connection:connection didReceiveData:data];
+}
+
+- (void)connection:(NSURLConnection*)connection didFailWithError:(NSError*)e {
+    EXPECT_TRUE([_operationQueue isEqual:[NSOperationQueue currentQueue]]);
+    [super connection:connection didFailWithError:e];
+}
+
 @end
 
 // Subclass of NSURLConnectionHelper that starts its connection on a separate thread's run loop.
@@ -122,6 +137,7 @@ typedef NS_ENUM(NSInteger, NSURLConnectionDelegateType) {
 @property (retain) NSThread* thread;
 - (void)doNothing;
 - (void)spinRunLoop;
+@property (assign) NSRunLoop* loop;
 @end
 
 @implementation NSURLConnectionTestHelper_RunLoop
@@ -145,14 +161,14 @@ typedef NS_ENUM(NSInteger, NSURLConnectionDelegateType) {
 // Keeps the run loop on the current thread spinning.
 - (void)spinRunLoop {
     @autoreleasepool {
-        NSRunLoop* loop = [NSRunLoop currentRunLoop];
+        _loop = [NSRunLoop currentRunLoop];
 
         NSTimer* timer = [NSTimer timerWithTimeInterval:0.1 target:self selector:@selector(doNothing) userInfo:nil repeats:YES];
-        [loop addTimer:timer forMode:NSDefaultRunLoopMode];
+        [_loop addTimer:timer forMode:NSDefaultRunLoopMode];
 
         while (!_isStopped) {
             @autoreleasepool {
-                [loop runMode:NSDefaultRunLoopMode beforeDate:[NSDate distantFuture]];
+                [_loop runMode:NSDefaultRunLoopMode beforeDate:[NSDate distantFuture]];
             }
         }
     }
@@ -163,6 +179,22 @@ typedef NS_ENUM(NSInteger, NSURLConnectionDelegateType) {
     [connection performSelector:@selector(start) onThread:_thread withObject:nil waitUntilDone:NO];
     return connection;
 }
+
+- (void)connection:(NSURLConnection*)connection didReceiveResponse:(NSURLResponse*)response {
+    EXPECT_TRUE([_loop isEqual:[NSRunLoop currentRunLoop]]);
+    [super connection:connection didReceiveResponse:response];
+}
+
+- (void)connection:(NSURLConnection*)connection didReceiveData:(nonnull NSData*)data {
+    EXPECT_TRUE([_loop isEqual:[NSRunLoop currentRunLoop]]);
+    [super connection:connection didReceiveData:data];
+}
+
+- (void)connection:(NSURLConnection*)connection didFailWithError:(NSError*)e {
+    EXPECT_TRUE([_loop isEqual:[NSRunLoop currentRunLoop]]);
+    [super connection:connection didFailWithError:e];
+}
+
 @end
 
 // Test to verify a request can be successfully made and a valid response and data is received.

--- a/tests/functionaltests/Tests/NSURLSession.mm
+++ b/tests/functionaltests/Tests/NSURLSession.mm
@@ -281,7 +281,6 @@ class NSURLSessionTests {
 public:
     BEGIN_TEST_CLASS(NSURLSessionTests)
     TEST_CLASS_PROPERTY(L"UAP:AppXManifest", L"NSURL.AppxManifest.xml")
-    TEST_CLASS_PROPERTY(L"ignore", L"true")
     END_TEST_CLASS()
 
     TEST_CLASS_SETUP(NSURLClassSetup) {

